### ISSUE TITLE
Wildcard Certificate Improvements

### ISSUE THREAD

**Issue by Stebalien** [mkg20001/libp2p-nodetrust#4]
Once lets encrypt supports wildcard certificates, consider:

1. Having everyone use the same wildcard cert + key.
2. Putting the key + cert in IPFS and creating a dnslink record (with a really long TTL) mapping `/ipns/your.domain` to the IPFS directory with the cert+key.
2. Configuring the DHS server to map IP.your.domain to IP (with a really long TTL).

This will:

1. Reduce the load on the DNS server (long TTLs).
2. Require *only* the special DNS server + an IPFS server.
3. Allow peers to use this service *without* having to contact any external service when they change their IP address.
4. Allow us to ship IPFS with built-in support for this system. The only necessary configuration will be some setting telling IPFS to use some set of SSL domains.

**Comment by mkg20001** [comment]
> Having everyone use the same wildcard cert + key.

Already does that

Now I finally get that idea.

> Putting the key + cert in IPFS and creating a dnslink record (with a really long TTL) mapping /ipns/your.domain to the IPFS directory with the **cert+key.**

I was using a libp2p node for that because putting it on ipfs wouldn't work for libp2p things that don't use ipfs but use libp2p.

**Comment by Stebalien** [comment]
> I was using a libp2p node for that because putting it on ipfs wouldn't work for libp2p things that don't use ipfs but use libp2p.

That's reasonable. However, if you *also* put it in IPFS, it would make it really easy for IPFS nodes to use this (and make it as decentralized as possible given the DNS and CA dependency).

**Comment by mkg20001** [comment]
@Stebalien But a libp2p node is still required for discovery, right?

**Comment by Stebalien** [comment]
Peer discovery? Yes.

My point is that this would allow us to easily use this from IPFS (go and js). Currently, one needs to stick an NGINX reverse proxy out in front of an IPFS node to get it to listen on an SSL enabled websocket (because all the necessary SSL setup/cert stuff). However, given a system like this, I believe we could get rid of that.

Given the following IPFS config snippet:

```
"Addresses": {
  "SslReflectors": [ "a.com", "b.com" ]`,
  "Swarm": [
    "/ip4/0.0.0.0/tcp/4001",
    "/ip4/0.0.0.0/tcp/443/wss"
]
```

IPFS would bootstrap and connect to the ipfs/libp2p network normally. Then, the WSS (secure websocket) transport would:

1. Resolve `/ipns/DOMAIN` for each domain in `Addresses.SslReflectors`.
2. Fetch the associated SSL certificate/keys through IPFS (no HTTP necessary).
3. Start listening on the specified `wss` addresses using the fetched SSL keys (using SNI to distinguish between the different domains?).

**Comment by mkg20001** [comment]
@Stebalien 10/10

Although it should also allow to provide local certificates because that would remove the need for a reverse proxy.

So something like:
```
"Addresses": {
  "SslConfig": {
    "cert": "cert.pem",
    "key": "privkey.pem"
  },
  "Swarm": [
    "/ip4/0.0.0.0/tcp/4001",
    "/ip4/0.0.0.0/tcp/443/wss"
]
```
should also be possible

**Comment by mkg20001** [comment]
or wait. even better. also includes nodes that don't have ipfs, only libp2p:
```js
{
  "Addresses": {
    "SslConfig": [ // use SNI to switch between domains
      {
        "type": "file",
        "cert": "example.com.cert.pem",
        "key": "example.com.key.pem"
      },
      {
        "type": "file",
        "cert": "example.tk.cert.pem",
        "key": "example.tk.key.pem"
      },
      {
        "type": "ipns-ssl-reflector",
        "ipns": "a.com"
      },
      {
        "type": "ipns-ssl-reflector",
        "ipns": "b.com"
      },
      {
        "type": "libp2p-ssl-reflector",
        "node": {
          "Id": "Qm...",
          "Addr": [
            "/dns4/example.de/tcp/443/wss/ipfs/Qm..."
          ]
        }
      }
    ],
    "Swarm": [
      "/ip4/0.0.0.0/tcp/4001",
      "/ip4/0.0.0.0/tcp/443/wss"
    ]
  }
}
```

**Comment by mkg20001** [comment]
Basically your idea requires this thing to be rewritten from scratch so I have created https://github.com/mkg20001/libp2p-ssl-reflector
The DNS part: https://github.com/mkg20001/libp2p-ssl-reflector/blob/master/server/dns/index.js#L8-L24

**Comment by mkg20001** [comment]
It's live: https://asciinema.org/a/TATQv9THtsk8zoSkoD5QsZ7tG

**Comment by Stebalien** [comment]
Awesome!

What's the difference between "ipns-ssl-reflector" and "libp2p-ssl-reflector"?

> Although it should also allow to provide local certificates because that would remove the need for a reverse proxy.

My idea was to fetch the certs through IPFS (although, yes, if you don't have IPFS you'd need to provide them locally). I agree, we shouldn't need a reverse proxy no matter how what we do.

**Comment by mkg20001** [comment]
> What's the difference between "ipns-ssl-reflector" and "libp2p-ssl-reflector"?

If a client uses only libp2p - without ipfs - then fetching the certs from ipns is not possible. That's basically a workaround using libp2p.

**Comment by mkg20001** [comment]
The fact that certificates can be revoked by anyone having the certificate and private key require a few changes to nodetrust:
 - Certs need to be obtained for every domain individually
 - The csr probably needs to be changed
 - The DNS provider stuff can be removed, leaving only a simple dns server behind

**Comment by mkg20001** [comment]
@Stebalien currently working on those changes in #5

**Comment by Stebalien** [comment]
Nice! Now if we could just get browsers to allow webapps to validate SSL certificates when connecting by IP address...

**Comment by mkg20001** [comment]
@Stebalien Actually we can use letsencrypt for that. The rate limits apply for every subdomain individually so nodetrust can issue certs for as many subdomains as needed! 🎉

**Comment by Stebalien** [comment]
Nice! But can't the subdomain ratelimit cause a problem for ephemeral IP addresses?

**Comment by mkg20001** [comment]
No, unless the client spams the cert endpoint (which should also be preventable by nodetrust itself)
Dynamic ips get re-assigned around every 24h and afaik 20 certs per week per domain are allowed
